File: orchestration-os-merged/scripts/nodejs/agent.js
#!/usr/bin/env node
/**
 * Node.js AI Agent - CLI Tool
 * Accepts natural language prompts and generates/executes JavaScript code
 */

const { execSync } = require('child_process');
const { writeFileSync, unlinkSync } = require('fs');
const https = require('https');
const http = require('http');

const OLLAMA_HOST = process.env['OLLAMA_HOST'] || 'http://localhost:11434';
const OLLAMA_MODEL = process.env['OLLAMA_MODEL'] || 'qwen2.5-coder:3b';
const GEMINI_API_KEY = process.env['GEMINI_API_KEY'] || '';
const OUTPUT_DIR = process.env['OUTPUT_DIR'] || '/tmp';

const COLOR = {
  red: '\x1b[0;31m',
  green: '\x1b[0;32m',
  yellow: '\x1b[1;33m',
  reset: '\x1b[0m'
};

function log(level, msg) {
  const prefix = level === 'info' ? `${COLOR.green}[Agent]${COLOR.reset}` :
                 level === 'error' ? `${COLOR.red}[Error]${COLOR.reset}` :
                 `${COLOR.yellow}[Warn]${COLOR.reset}`;
  console.error(`${prefix} ${msg}`);
}

function fetchJSON(url, options = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const client = urlObj.protocol === 'https:' ? https : http;
    const req = client.request(url, options, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try { resolve(JSON.parse(data)); }
        catch { reject(new Error(`Invalid JSON from ${url}`)); }
      });
    });
    req.on('error', reject);
    req.end();
  });
}

async function callOllama(prompt) {
  try {
    const body = JSON.stringify({
      model: OLLAMA_MODEL,
      messages: [{ role: 'user', content: prompt }],
      stream: false
    });
    const result = await fetchJSON(OLLAMA_HOST + '/api/chat', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body
    });
    return result?.message?.content ?? null;
  } catch (e) {
    log('warn', `Ollama call failed: ${e.message}`);
    return null;
  }
}

async function callGemini(prompt) {
  if (!GEMINI_API_KEY) return null;
  try {
    const body = JSON.stringify({
      contents: [{ parts: [{ text: prompt }] }],
      generationConfig: { temperature: 0.7, maxOutputTokens: 8192 }
    });
    const result = await fetchJSON(
      `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=${GEMINI_API_KEY}`,
      { method: 'POST', headers: { 'Content-Type': 'application/json' }, body }
    );
    return result?.candidates?.[0]?.content?.parts?.[0]?.text ?? null;
  } catch (e) {
    log('warn', `Gemini call failed: ${e.message}`);
    return null;
  }
}

function cleanCode(raw) {
  return raw
    .replace(/```(?:javascript|js|typescript|ts|node)?/gi, '')
    .replace(/```/g, '')
    .replace(/`/g, '')
    .trim();
}

async function generateCode(prompt) {
  const fullPrompt = `Generate ONLY JavaScript/Node.js code for: ${prompt}\n\nRules:
- Output ONLY the code, no explanations
- No markdown code blocks
- Code must be valid Node.js/JavaScript
- Include proper require() statements
- No bash commands or shell script`;

  let result = await callOllama(fullPrompt);
  if (!result) result = await callGemini(fullPrompt);
  if (!result) return null;

  return cleanCode(result);
}

async function executeCode(code) {
  const tmpFile = `${OUTPUT_DIR}/agent_${Date.now()}.js`;
  writeFileSync(tmpFile, code);

  try {
    const output = execSync(`node "${tmpFile}"`, { encoding: 'utf8', timeout: 60000 });
    return { output, exitCode: 0 };
  } catch (e) {
    return {
      output: (e.stderr || String(e)),
      exitCode: e.status ?? 1
    };
  } finally {
    try { unlinkSync(tmpFile); } catch {}
  }
}

async function testConnection() {
  log('info', 'Testing LLM connectivity...');

  const ollamaResult = await callOllama('Say "OK" in exactly one word');
  if (ollamaResult && ollamaResult.toLowerCase().includes('ok')) {
    log('info', 'Ollama connection: OK');
  } else {
    log('error', 'Ollama connection: FAILED');
  }

  if (GEMINI_API_KEY) {
    const geminiResult = await callGemini('Say "OK" in exactly one word');
    if (geminiResult && geminiResult.toLowerCase().includes('ok')) {
      log('info', 'Gemini connection: OK');
    } else {
      log('error', 'Gemini connection: FAILED');
    }
  } else {
    log('info', 'Gemini: Not configured');
  }
}

function showHelp() {
  console.log(`Node.js AI Agent - Generate and execute JavaScript using AI

Usage: node agent.js <prompt>
       node agent.js --test

Examples:
  node agent.js "Read a JSON file and print its keys"
  node agent.js "Fetch JSON from an API and display it"
  node agent.js --test

Environment Variables:
  OLLAMA_HOST      Ollama server URL (default: http://localhost:11434)
  OLLAMA_MODEL     Model to use (default: qwen2.5-coder:3b)
  GEMINI_API_KEY   Google Gemini API key (optional)
  OUTPUT_DIR       Temp directory for code execution (default: /tmp)
`);
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    showHelp();
    process.exit(1);
  }

  if (args[0] === '--test') {
    await testConnection();
    return;
  }

  if (args[0] === '--help' || args[0] === '-h') {
    showHelp();
    return;
  }

  const prompt = args.join(' ');
  log('info', `Generating code for: ${prompt}`);

  const code = await generateCode(prompt);
  if (!code) {
    log('error', 'No LLM available and no code generated');
    process.exit(1);
  }

  log('info', `Generated ${code.split('\n').length} lines of code`);
  console.log(`\n--- Generated Code ---\n${code}\n--- End Code ---\n`);

  log('info', 'Executing...');
  const { output, exitCode } = await executeCode(code);

  console.log(output);
  process.exit(exitCode);
}

main().catch(e => {
  log('error', `Fatal: ${e.message}`);
  process.exit(1);
});

File: orchestration-os-merged/scripts/go/agent.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"time"
)

var (
	ollamaHost  = getEnv("OLLAMA_HOST", "http://localhost:11434")
	ollamaModel = getEnv("OLLAMA_MODEL", "qwen2.5-coder:3b")
	geminiAPIKey = os.Getenv("GEMINI_API_KEY")
)

type OllamaRequest struct {
	Model    string                   `json:"model"`
	Messages []map[string]interface{} `json:"messages"`
	Stream   bool                     `json:"stream"`
}

type OllamaResponse struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
}

type GeminiRequest struct {
	Contents []map[string]interface{} `json:"contents"`
}

func getEnv(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}

func callOllama(prompt string) (string, error) {
	reqBody := OllamaRequest{
		Model: ollamaModel,
		Messages: []map[string]interface{}{
			{"role": "user", "content": prompt},
		},
		Stream: false,
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", err
	}

	client := &http.Client{Timeout: 120 * time.Second}
	req, err := http.NewRequest("POST", ollamaHost+"/api/chat", bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result OllamaResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	return result.Message.Content, nil
}

func callGemini(prompt string) (string, error) {
	if geminiAPIKey == "" {
		return "", fmt.Errorf("GEMINI_API_KEY not set")
	}

	reqBody := GeminiRequest{
		Contents: []map[string]interface{}{
			{"parts": []map[string]string{{"text": prompt}}},
		},
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", err
	}

	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=%s", geminiAPIKey)
	client := &http.Client{Timeout: 60 * time.Second}
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	candidates, ok := result["candidates"].([]interface{})
	if !ok || len(candidates) == 0 {
		return "", fmt.Errorf("no candidates in response")
	}

	candidate, ok := candidates[0].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid candidate format")
	}

	content, ok := candidate["content"].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("no content in candidate")
	}

	parts, ok := content["parts"].([]interface{})
	if !ok || len(parts) == 0 {
		return "", fmt.Errorf("no parts in content")
	}

	part, ok := parts[0].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid part format")
	}

	text, ok := part["text"].(string)
	if !ok {
		return "", fmt.Errorf("no text in part")
	}

	return text, nil
}

func generateCode(prompt string) (string, error) {
	fullPrompt := fmt.Sprintf(`You are an expert Go programmer. Generate ONLY Go code for this request:

%s

RULES:
- Output ONLY Go code, nothing else
- NO explanations, NO comments describing what the code does
- NO markdown formatting, NO backticks, NO code block markers
- Code must be valid Go syntax
- Code must be complete and runnable
- Use proper error handling with if err != nil
- Use fmt for output (fmt.Println, fmt.Printf, etc.)
- Include proper imports

Output just the raw Go code now:`, prompt)

	if result, err := callOllama(fullPrompt); err == nil && result != "" {
		return cleanCode(result), nil
	}

	if result, err := callGemini(fullPrompt); err == nil && result != "" {
		return cleanCode(result), nil
	}

	return "", fmt.Errorf("no LLM available - configure OLLAMA_HOST or GEMINI_API_KEY")
}

func cleanCode(raw string) string {
	re := regexp.MustCompile("```go")
	raw = re.ReplaceAllString(raw, "")
	re = regexp.MustCompile("```")
	raw = re.ReplaceAllString(raw, "")
	raw = strings.ReplaceAll(raw, "`", "")
	return strings.TrimSpace(raw)
}

func executeCode(code string) (string, string, int) {
	tmpFile, err := os.CreateTemp("", "agent_*.go")
	if err != nil {
		return "", fmt.Sprintf("Failed to create temp file: %v", err), 1
	}
	defer os.Remove(tmpFile.Name())
	defer tmpFile.Close()

	if _, err := tmpFile.WriteString(code); err != nil {
		return "", fmt.Sprintf("Failed to write code: %v", err), 1
	}
	tmpFile.Close()

	var stdout, stderr bytes.Buffer
	cmd := exec.Command("go", "run", tmpFile.Name())
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err = cmd.Run()
	exitCode := 0
	if err != nil {
		exitCode = 1
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		}
	}

	return stdout.String(), stderr.String(), exitCode
}

func showHelp() {
	fmt.Println(`Go AI Agent - Generate and execute Go code using AI

Usage: agent <prompt>
       agent --test

Examples:
  agent "Print hello world 10 times"
  agent "Read a file and count its lines"
  agent --test

Environment Variables:
  OLLAMA_HOST      Ollama server URL (default: http://localhost:11434)
  OLLAMA_MODEL     Model to use (default: qwen2.5-coder:3b)
  GEMINI_API_KEY   Google Gemini API key (optional)
`)
}

func main() {
	if len(os.Args) < 2 {
		showHelp()
		os.Exit(1)
	}

	arg := os.Args[1]

	if arg == "--test" {
		fmt.Fprintln(os.Stderr, "[Agent] Testing LLM connectivity...")

		if _, err := callOllama("Say 'Hello' in exactly one word"); err != nil {
			fmt.Fprintf(os.Stderr, "[Error] Ollama connection: FAILED (%v)\n", err)
		} else {
			fmt.Fprintln(os.Stderr, "[Agent] Ollama connection: OK")
		}

		if geminiAPIKey != "" {
			if _, err := callGemini("Say 'Hello' in exactly one word"); err != nil {
				fmt.Fprintf(os.Stderr, "[Error] Gemini connection: FAILED (%v)\n", err)
			} else {
				fmt.Fprintln(os.Stderr, "[Agent] Gemini connection: OK")
			}
		} else {
			fmt.Fprintln(os.Stderr, "[Agent] Gemini: Not configured")
		}
		os.Exit(0)
	}

	prompt := strings.Join(os.Args[1:], " ")
	fmt.Fprintf(os.Stderr, "[Agent] Generating code for: %s\n", prompt)

	code, err := generateCode(prompt)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[Agent Error] %v\n", err)
		os.Exit(1)
	}

	fmt.Fprintf(os.Stderr, "[Agent] Generated %d chars of code\n", len(code))

	stdout, stderr, exitCode := executeCode(code)

	if stdout != "" {
		fmt.Print(stdout)
	}
	if stderr != "" {
		fmt.Fprintf(os.Stderr, "[Error] %s\n", stderr)
	}

	os.Exit(exitCode)
}
